assets/cmo_barcodes.csv

```diff
@@ -0,0 +1,3 @@
+sample_id,cmo_ids,description
+PBMCs_human_1,CMO301,PBMCs_human_1
```

Comment:
Hi @grst ,
I cannot give much information. This is something I got from the `cellranger/multi` module testings and all, and there I think it was already different (so I did not change).

Hi @klkeys, Could you shed some light on this?

Should indeed we make `sample_id` the same of the one in the `samplesheet` (which I agree with @grst is the logical thinking) or this is indeed something different?

Reply:
actually, I think I answered it below

Reply:
I don't think that physical sample IDs need to match the CMO IDs, see the [CMO test data](https://github.com/nf-core/test-datasets/blob/modules/data/genomics/homo_sapiens/10xgenomics/cellranger/10k_pbmc_cmo/README.md) for the `cellranger multi` module

matching physical and CMO IDs does sense if you have exactly one CMO sample per physical sample, as in @grst's explanation below

note that the CMO Feature Reference CSV only needs controls, and it requires all controls from the same CMO sample as one line of the `cellranger multi` config ([doc reference](https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/running-pipelines/cr-3p-multi#multi-gex-cellplex-more))

Reply:
also relevant to https://github.com/nf-core/scrnaseq/pull/276#discussion_r1551764285

I haven't run CMO myself, but I don't think that you tag multiple samples with the same CMO

your CMO files should be this:

**proposed CMO samplesheet**
```
sample,multiplexed_sample_id,probe_barcode_ids,description
SAMP001,SAMP001_ctrl,BC301,Control
SAMP001,SAMP001_trt,BC302,Treated
SAMP002,SAMP002_ctrl,BC303,Control
SAMP002,SAMP002_trt,BC304,Treated
```

**sample 1 CMO config***
```
sample_id,probe_barcode_ids,description
SAMP001_ctrl,BC301,Control
SAMP001_ctrl,BC301,Control
```

**sample 2 CMO config***
```
sample_id,probe_barcode_ids,description
SAMP002_ctrl,BC303,Control
```

Reply:
Okay. Thanks both. I am resolving this one, so it can be taken all together in the ""main"" comment related to it.

Reply:
Hi both,

So, I have come up with a way to have such a "unified" samplesheet and then, perform the parsing for CMO and FRNA information from a ""unified"" cellranger multi barcodes samplesheet.

Basically, this is the code that allows it, in the `align_cellrangermulti.nf` subworkflow:

```groovy
        // parse frna and barcode information
        if (params.cellranger_multi_barcodes) {
            Channel.fromPath( params.cellranger_multi_barcodes )
            .splitCsv( header: true )
            .map {
                assert it.containsKey('sample'), "The ${params.cellranger_multi_barcodes} samplesheet for cellranger multi must contain a 'sample' colum. Please, check the docs."
                assert it.containsKey('multiplexed_sample_id'), "The ${params.cellranger_multi_barcodes} samplesheet for cellranger multi must contain a 'multiplexed_sample_id' colum. Please, check the docs."
                assert it.containsKey('description'), "The ${params.cellranger_multi_barcodes} samplesheet for cellranger multi must contain a 'description' colum. Please, check the docs."
                assert it.containsKey('probe_barcode_ids') || it.containsKey('cmo_ids'), "The ${params.cellranger_multi_barcodes} samplesheet for cellranger multi must contain at least one, or both, of 'cmo_ids' or 'probe_barcode_ids' columns. Please, check the docs."

                it
            }
            .branch {
                cmo  : it.containsKey('cmo_ids')
                frna : it.containsKey('probe_barcode_ids')
            }
            .set { ch_cellrangermulti_barcodes }

            //
            // Here, we parse the received cellranger multi barcodes samplesheet.
            // We get the sample name information and merge the "sample-map" with the received GEX fastqs.
            // The selection of the GEX fastqs is because samples are always expected to have at least GEX data.
            // Then, using "joined" map, which means, the "additional barcode information" of each sample, we then,
            // parse it to generate the cmo / frna samplesheets to be used by each sample.
            //
            // Here, because of the .join() we take advantage of the "FIFO"-rule and are sure that the data used in the
            // module is from the same sample from the "normal" samplesheet.
            //

            // CMOs
            ch_cmo_barcode_csv =
            ch_grouped_fastq.gex.map{ it[0].id }
            .join( ch_cellrangermulti_barcodes.cmo.collect().map{ [ it[0].sample, it ] }.groupTuple(by:0).ifEmpty{ [ [], [] ] }, remainder: true ).filter{ it[0] != [] } // .ifEmpty{} is there to allow join when data is not available and the rest do not fail, and the . filter{} is there to remove the generated [[],[]] when empty so it does not break the expected channel size.
            .map{ sample, meta ->
                // groupTuple adds one level nest on array, thus, meta[0] takes the real array with all maps.
                if (meta && meta[0][0].containsKey('cmo_ids')) {
                    lines = [ 'sample_id,cmo_ids,description' ]
                    meta[0].each{ lines = lines + [ "$it.multiplexed_sample_id,$it.cmo_ids,$it.description" ] }
                    cmos = file( "${sample}_cmo_samplesheet.csv" )
                    cmos.text = lines.join("\n").trim().toString()
                    cmos
                } else { empty_file }
            }

            // FRNAs
            ch_frna_sample_csv =
            ch_grouped_fastq.gex.map{ it[0].id }
            .join( ch_cellrangermulti_barcodes.frna.collect().map{ [ it[0].sample, it ] }.groupTuple(by:0).ifEmpty{ [ [], [] ] }, remainder: true ).filter{ it[0] != [] } // .ifEmpty{} is there to allow join when data is not available and the rest do not fail, and the . filter{} is there to remove the generated [[],[]] when empty so it does not break the expected channel size.
            .map{ sample, meta ->
                // groupTuple adds one level nest on array, thus, meta[0] takes the real array with all maps.
                if (meta && meta[0][0].containsKey('probe_barcode_ids')) {
                    lines = [ 'sample_id,probe_barcode_ids,description' ]
                    meta[0].each{ lines = lines + [ "$it.multiplexed_sample_id,$it.probe_barcode_ids,$it.description" ] }
                    frna = file( "${sample}_frna_samplesheet.csv" )
                    frna.text = lines.join("\n").trim().toString()
                    frna
                } else { empty_file }
            }

        } else {
            ch_cmo_barcode_csv = empty_file
            ch_frna_sample_csv = empty_file
        }
```

The testing dataset only allows me to test the execution with CMOs. Can you two take a look at it, and provide some reviewing feedback? Would also be nice to test it out with some data you may have.

Execution was:

```console
nextflow run ../main.nf -profile test_cellrangermulti,docker
```

Test config:

```groovy
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Nextflow config file for running minimal tests
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Defines input files and everything required to run a fast and simple pipeline test.

    Use as follows:
        nextflow run nf-core/scrnaseq -profile test,<docker/singularity> --outdir <OUTDIR>

----------------------------------------------------------------------------------------
*/

// shared across profiles
params {
    config_profile_name        = 'Test profile (Cellranger Multi)'
    config_profile_description = 'Minimal test dataset to check pipeline function using cellranger multi'

    // Resources on test case
    max_cpus   = 10
    max_memory = '50.GB'
    max_time   = '6.h'

    // Input data
    input                     = "${projectDir}/assets/cellrangermulti_samplesheet.csv"
    cellranger_multi_barcodes = "${projectDir}/assets/cellranger_barcodes_samplesheet.csv"
    skip_emptydrops           = true // not enough data in small test

    // Genome references
    fasta           = 'https://ftp.ensembl.org/pub/release-110/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz'
    gtf             = 'https://ftp.ensembl.org/pub/release-110/gtf/homo_sapiens/Homo_sapiens.GRCh38.110.gtf.gz'

    // aligner
    aligner         = 'cellrangermulti'
    protocol        = 'auto'

    // other
    validationSchemaIgnoreParams = 'genomes'
}

```

Reply:
thanks, from reading the code the logic seems to make sense to me. 
I only have a multiplexed FFPE dataset I could test on, I'll try to do so this week.

Reply:
thanks, from reading the code the logic seems to make sense to me. 
I only have a multiplexed FFPE dataset I could test on, I'll try to do so this week.

Reply:
I am fixing some stuff (which I will update when done) then you can test it afterwards.

Reply:
Hi @grst ,
I have added the proposed changes and other fixes that I found were around:

**Code as a module**
First of all, as suggested [here](https://github.com/nf-core/scrnaseq/pull/276#discussion_r1559315917) I have modified the parsing code from Groovy to a Module so we face no problems with AWS neither with `caching` feature.

The module is a python script that checks the consistency of the additional samplesheet and, based on the `cmo / frna` related columns, it splits the samplesheet having one for each sample.

**Adding module splitted samplesheets to workflow**
Overall, the code above is simple. The tricky part was adding it to the context of the code again. Here, I took advantage of the "FIFO" rules and using the "GEX" files channel as a base to re-connect and order the split `cmo / frna` samplesheets generated so that they are used, in the same order (the correct samples) in the `CELLRANGER_MULTI` module.

This is done in this chunk:
https://github.com/nf-core/scrnaseq/blob/247-support-for-10x-ffpe-scrna/subworkflows/local/align_cellrangermulti.nf#L64-L105

**Parsing the generated results for `MTX_CONVERSION`**
Then, I had to add a parsing for the generated results, in order to be able to convert the generated data to `.h5ad / .Rds`. Cellranger multi, outputs the filtered results in a special folder called `per_sample_outs` so that, when you have multiple samples demultiplexed by the barcodes given, they will be each in a subdir there.

As such, when doing `cellranger/multi`, we will be converting the `raw_matrices` ( not per sample ) and the `filtered_matrices` (per sample).

The code related to it is here:
https://github.com/nf-core/scrnaseq/blob/247-support-for-10x-ffpe-scrna/subworkflows/local/align_cellrangermulti.nf#L181-L228

The final part of it is just the "standard" splitting raw / filtered as we do for normal cellranger as well.

**Custom Emptydrops**
Currently, emptydrops will not be performed for `cellranger/multi` as I am not sure if it is relevant.
https://github.com/nf-core/scrnaseq/blob/247-support-for-10x-ffpe-scrna/workflows/scrnaseq.nf#L278

**Samplesheet parsing**
Because the cellranger multi subworkflow can receive data from multiple feature types for the same sample, but we must preserve the features type per sample information in different channels so they can be properly parsed [here](https://github.com/nf-core/scrnaseq/blob/247-support-for-10x-ffpe-scrna/workflows/scrnaseq.nf#L223-L257) to create the correct channels expected modules.

So, I had to adapt the samplesheet parsing here: https://github.com/nf-core/scrnaseq/blob/247-support-for-10x-ffpe-scrna/subworkflows/local/utils_nfcore_scrnaseq_pipeline/main.nf#L83-L124

**nf-tests and documentation**
Nf-tests and documentation need to be worked on. But I will only start it once we resolve the code.

😄

Reply:
Now it can be tested with your data so you can also see how the outputs look like.

Reply:
Fantastic! I'll try it on Monday, have a nice weekend!